torchao/experimental/quant_api.py

```diff
@@ -463,3 +463,106 @@ def quantize(self, model: nn.Module) -> nn.Module:
             },
         )
         return model
+
+
+class IntxWeightOnlyQuantizedLinear(nn.Module):
+    def __init__(
+        self,
+        pack_weight_op,
+        linear_op,
+    ):
+        super().__init__()
+        self._pack_weights_op = pack_weight_op
+        self._linear_op = linear_op
+
+    def quantize_and_pack_weights(self, weights, nbit, group_size):
+        self.nbit = nbit
+        self.group_size = group_size
+
+        weight_qvals, weight_scales, weight_zeros = _quantize(
+            weights, self.group_size, self.nbit, True
+        )
+        weight_qvals = (weight_qvals + (1 << (nbit - 1))).to(torch.uint8)
+
+        self.weight_scales_and_zeros = torch.stack(
+            (weight_scales.t(), weight_zeros.t()), dim=2
+        )
+
+        self.packed_weights = self._pack_weights_op(weight_qvals.cpu()).to(device="mps")
+
+    def forward(self, x):
+        assert x.dim() == 2
```

Comment:
Should we make it work for arbitrary shape to match behavior of torch.linear?  See cpu impls for reshaping.

Reply:
done!